proxysrv/server.go

```diff
@@ -0,0 +1,330 @@
+}
+
+// Option modifies an internal options type.
+type Option func(*options)
+
+type options struct {
+	log            *logrus.Entry
+	forwardTimeout time.Duration
+	traceClient    *trace.Client
+}
+
+// New creates a new Server with the provided destinations. The server returned
+// is unstarted.
+func New(destinations *consistent.Consistent, opts ...Option) (*Server, error) {
+	res := &Server{
+		Server: grpc.NewServer(),
+		opts:   &options{},
+		conns:  newClientConnMap(grpc.WithInsecure()),
+	}
+
+	for _, opt := range opts {
+		opt(res.opts)
+	}
+
+	if res.opts.log == nil {
+		log := logrus.New()
+		log.Out = ioutil.Discard
+		res.opts.log = logrus.NewEntry(log)
+	}
+
+	if err := res.SetDestinations(destinations); err != nil {
+		return nil, fmt.Errorf("failed to set the destinations: %v", err)
+	}
+
+	forwardrpc.RegisterForwardServer(res.Server, res)
+
+	return res, nil
+}
+
+// Serve starts a gRPC listener on the specified address and blocks while
+// listening for requests. If listening is iterrupted by some means other than
+// Stop or GracefulStop being called, it returns a non-nil error.
+func (s *Server) Serve(addr string) (err error) {
+	ln, err := net.Listen("tcp", addr)
+	if err != nil {
+		return fmt.Errorf("failed to bind the proxy server to '%s': %v",
+			addr, err)
+	}
+	defer func() {
+		if cerr := ln.Close(); err == nil && cerr != nil {
+			err = cerr
+		}
+	}()
+
+	return s.Server.Serve(ln)
+}
+
+// Stop stops the gRPC server (if it was started) and closes all gRPC client
+// connections
+func (s *Server) Stop() {
+	s.Server.Stop()
+	s.conns.Clear()
+}
+
+// SetDestinations updates the ring of hosts that are forwarded to by
+// the server.  If new hosts are being added, a gRPC connection is initialized
+// for each.
+//
+// This also prunes the list of open connections.  If a connection exists to
+// a host that wasn't in either the current list or the last one, the
+// connection is closed.
+func (s *Server) SetDestinations(dests *consistent.Consistent) error {
+	s.updateMtx.Lock()
+	defer s.updateMtx.Unlock()
+
+	var current []string
+	if s.destinations != nil {
+		current = s.destinations.Members()
+	}
+	new := dests.Members()
+
+	// for every connection in the map that isn't in either the current or
+	// previous list of destinations, delete it
+	for _, k := range s.conns.Keys() {
+		if !strInSlice(k, current) && !strInSlice(k, new) {
+			s.conns.Delete(k)
+		}
+	}
+
+	// create a connection for each destination
+	for _, dest := range new {
+		if err := s.conns.Add(dest); err != nil {
+			return fmt.Errorf("failed to setup a connection for the "+
+				"destination '%s': %v", dest, err)
+		}
+	}
+
+	s.destinations = dests
+	return nil
+}
+
+// SendMetrics spawns a new goroutine that forwards metrics to the destinations
+// and exist immediately.
+func (s *Server) SendMetrics(ctx context.Context, mlist *forwardrpc.MetricList) (*empty.Empty, error) {
+	go func() {
+		_ = s.sendMetrics(context.Background(), mlist)
```

Comment:
That seems totally reasonable to me.  The reason I implemented the gRPC handler this way was to mirror the HTTP handler, which just parses the JSON body of the request and then spawns a new goroutine to handle proxying the metrics to the proper locations.

I wonder if there's any reason why this couldn't just block while the metrics are proxied (effectively removing the goroutine created here)?  Flushing on the side of a local veneur is not a blocking operation (it also spawns a new goroutine for just the forwarding process), so I can't really see a downside of this approach.  Making this RPC block would also let us return a proper error to the local Veneur if any part of the path to the final aggregating Veneur failed, which I think would be a really nice property.  As is, I've had a hard time finding a way to measure the number of metrics forwarded from a local Veneur that successfully make it to a global Veneur.

What do you think?

Reply:
Oh nooo, I missed that - sorry, thought this introduced an additional break in the context chain! This looks reasonable & all your suggestions seem ok also, though we wouldn't be able to cancel the timeout (since this function here terminates immediately after spawning the goroutines).

Reply:
@asf-stripe That does make sense, I suppose we'll need to leave the cancelling in the forked `sendMetrics` function.  And no worries 😄 

I'll look at making a default timeout for `sendMetrics`, and adding some instrumentation on the number of active proxying goroutines.

Reply:
@asf-stripe @sdboyer-stripe I updated this in faea0102 to include an atomic counter tracking the number of running goroutines.  I also added a new goroutine to the `(*proxysrv.Server).Serve` method that periodically reports this value (every 10 seconds).

I also added a default forwarding timeout of 10 seconds, which I just picked pretty arbitrarily.